### PR TITLE
ASoC: SOF: ipc3: Remove the ipc3-ops.h header file

### PR DESCRIPTION
**Pull request by ujfalusi**
Only the main IPC ops struct should be visible outside of IPC3 code to make
sure that the code is correctly abstracted.

Instead of keeping the ipc3-ops.h with only the high level ops struct
declaration, put the ipc3_ops to sof-priv.h and move all other ops struct
declaration into ipc3-priv.h

New IPC implementation should follow this route: the main IPC ops should be
declared in sof-priv.h and no other IPC version related header be used
for generic code.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

### CODE REVIEW

**Review by plbossart — APPROVED** [review]

**Comment by ujfalusi** [comment]
@plbossart, @ranj063, Would it be better to add the main IPC ops to some other (new) header file which is not included to virtually all SOF c file, like the sof-priv.h?

In practice this should be only used by the ipc.c after all.

Just a thought.

**Comment by plbossart** [comment]
> @plbossart, @ranj063, Would it be better to add the main IPC ops to some other (new) header file which is not included to virtually all SOF c file, like the sof-priv.h?
> 
> In practice this should be only used by the ipc.c after all.
> 
> Just a thought.

it's true that including sof-priv.h everywhere isn't that great in terms of separation.

**Comment by ranj063** [comment]
@ujfalusi I like this change but for IPC4, there's one place we need ipc4-ops.h outside of the ipc4 code. https://github.com/thesofproject/linux/commit/17101588b5e0375aa01995fbedcefb0b88049e3a
How would we deal with it?

**Comment by ranj063** [comment]
> @ujfalusi I like this change but for IPC4, there's one place we need ipc4-ops.h outside of the ipc4 code. [1710158](https://github.com/thesofproject/linux/commit/17101588b5e0375aa01995fbedcefb0b88049e3a) How would we deal with it?

maybe we need to declare sof_ipc4_data in sof-priv just like ipc4_ops as well?

**Comment by ranj063** [comment]
> @plbossart, @ranj063, Would it be better to add the main IPC ops to some other (new) header file which is not included to virtually all SOF c file, like the sof-priv.h?

yes, makes sense

**Comment by ujfalusi** [comment]
> > @ujfalusi I like this change but for IPC4, there's one place we need ipc4-ops.h outside of the ipc4 code. [1710158](https://github.com/thesofproject/linux/commit/17101588b5e0375aa01995fbedcefb0b88049e3a) How would we deal with it?
> 
> maybe we need to declare sof_ipc4_data in sof-priv just like ipc4_ops as well?

@ranj063, hrm. `struct sof_ipc4_data` is ipc4 private data, no? if it is then it is declared in ipc4-priv.h and files which deals with IPC specific operations should be allowed to include the needed sound/soc/sof/ipcX-priv.h for their needs.

We need to put the main ops declarations with the exports in a header file to avoid sparse warnings. At the same time we would want to limit the exposure, because we care.
What about a `ipc-ops.h` or even an `ipc-priv.h` for these sort of things?

I' also cooking a Kconfig based IPC version opt in method so ARM guys can opt to not compile any IPC4 code, just IPC3 to cut down the size.
Basically platform can select which IPC versions they need, can support. Intel will say IPC3 and IPC4, AMD, NXP, Mediatek would only select IPC3.

**Comment by ranj063** [comment]
> @ranj063, hrm. `struct sof_ipc4_data` is ipc4 private data, no? if it is then it is declared in ipc4-priv.h and files which deals with IPC specific operations should be allowed to include the needed sound/soc/sof/ipcX-priv.h for their needs.

OK fine then.we'll include ipc4-priv.h in the platform files that need to set the IPC-specific ops.

**Review by ranj063 — APPROVED** [review]

**Review by bardliao — APPROVED** [review]

**Comment by ujfalusi** [comment]
Let's go with this patch for now and add the main IPC ops declaration to `sof-priv.h` for now. We should do header splitting at a later time and not rush it to end up with something which does not scale.
